### PR TITLE
feat: Add bracketing to lexer

### PR DESCRIPTION
**Pull request by croyzor**
Revived from #32. I was mistaken to close it.

### CODE REVIEW

**Review by acl-cqc — COMMENTED** [review]
Mostly this is nice - `Bracket.hs` does feel a little overblown, but generally it seems much nicer/neater than the last time we had bracketing done earlier, and there are some big improvements in error messages etc. too :).

I'm not quite gonna approve though until you can refactor `within` such that I understand it ;-) (and maybe improve `brackets`(`Worker`) :-) )

File: brat/Brat/Lexer/Bracketed.hs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+openCloseMismatchErr :: (FC, BracketType) -> (FC, BracketType) -> Error
+openCloseMismatchErr open (fcClose, bClose)
+  = Err (Just fcClose) (BracketErr (OpenCloseMismatch open bClose))
+
+unexpectedCloseErr :: FC -> BracketType -> Error
+unexpectedCloseErr fc b = Err (Just fc) (BracketErr (UnexpectedClose b))
+
+-- Parse between two brackets of the same type
+within :: (FC, BracketType) -- The nearest opening bracket to the left of us
```
Comment (acl-cqc) [inline brat/Brat/Lexer/Bracketed.hs:77]: Thanks - the comments here really helped and now I understand how it works ;-). See #73 !

**Review by acl-cqc — APPROVED** [review]
Looks good to me although please have a think about these 3 comments

**Comment by acl-cqc** [comment]
And, thanks @croyzor - nice work! :-)

File: brat/Brat/Lexer/Bracketed.hs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      in  skipChars (i - 1) (bts ++ [FlatTok (Token closeFC (closeTok b))] ++ rest)
+    skipChars i (FlatTok t:rest)
+     | i >= tokenLen t = skipChars (i - tokenLen t) rest
+     | otherwise = (start (fc t), FlatTok t:rest)
+
+    closeTok Paren = RParen
+    closeTok Square = RSquare
+    closeTok Brace = RBrace
+
+eofErr :: FC -> BracketType -> Error
```
[the remark this answers is not in the document]
Comment (croyzor) [reply brat/Brat/Lexer/Bracketed.hs:67]: I don't think I'm really getting it, but have uncurried anyway

**Comment by croyzor** [comment]
Re: `Show BToken` - this exists so that we can use it in the implementation of `VisualStream` which we need to pack up error messages but afaict, we don't actually call `showTokens` in our tests. It *might* come up when using the `dbg` debugging function from the parser library? So something vaguely sensible would be nice

**Pull request merged**